src/Symfony/Bundle/SecurityBundle/Security/FirewallMap.php

```diff
@@ -75,4 +78,34 @@ private function getFirewallContext(Request $request)
             }
         }
     }
+
+    /**
+     * @param FinishRequestEvent $event
+     */
+    public function onKernelFinishRequest(FinishRequestEvent $event)
+    {
+        $this->detachListeners($event->getRequest());
+    }
+
+    /**
+     * Cleans up the internal state of the firewall map.
+     *
+     * @param Request $request
+     */
+    private function detachListeners(Request $request)
+    {
+        unset($this->contexts[$request]);
+    }
```

Comment:
IMO you should directly unset the request in the `onKernelFinishRequest` method, that would make one less method call...